File: GCons/DP2.m
classdef DP2
    %dp2 constraint using r-p formulation. The dp2 constraint exists
    %between a vector aiBar on body i, and a vector dij which is the distance 
    %between point Pi and Point Qj, represented in the G-RF. dp2
    %constraint specifies that the dot product of these two vectors assume a
    %certian value, F(t), which is frequently a constant
    
    properties
        bodyi; %instance of a body object, i 
        bodyj; %instance of a body object, j
        
        %these point definitions are taken from S13 lecture 9.26
        Pi; %index of a [3x1] marker contained within bodyi, representing head of SiBar
        Qj; %index of a [3x1] marker contained within bodyj, representing head of SjBar
        ai; %index of a [3x1] marker contained within bodyi, representing the head of AiBar
        f; %prescribed value that the dot product should assume
        fdot; %derivative of above
        fddot;%derivative of above 
        t;  
    end
    
    %calculated values 
    properties(Dependent)
        aiBar; %vector AiBar the first vector in the DP2 constraint
        dij;   %distance in G-RF from point Pi to point Qj, the second vector in the DP2 constraint
        dijdot;%derivative of dij
        phi; %jacobian
        nu;  % dPhi_dp2/ dt
        gamma; %RHS of acceleration equation
        phi_ri;%derivative of phi wrt ri
        phi_rj;%derivative of phi wrt rj
        phi_pi;%derivative of phi wrt ri
        phi_pj;%derivative of phi wrt ri
        
        
        
    end
    
    methods
        %constructor
        function dp2 = DP2(bodyi,bodyj,Pi,Qj,ai,f,fdot,fddot,t)
            dp2.bodyi = bodyi;
            dp2.bodyj = bodyj;
            dp2.Pi = Pi;
            dp2.Qj = Qj;
            dp2.ai = ai;
            dp2.f = f;
            dp2.fdot = fdot;
            dp2.fddot = fddot;
            dp2.t = t;
            
        end
        %getters for dependent properties
        
        %consider revision that makes dij an abstract method
          function dij = get.dij(dp2)
            %returns the distance vector [3x1] in G-RF that connects point
            %Pi on body i to point Qj on body j
            dij = dp2.bodyj.r + dp2.bodyj.A*dp2.bodyj.markers{dp2.Qj} - ...
                 (dp2.bodyi.r - dp2.bodyi.A*dp2.bodyi.markers{dp2.Pi});
        end
        
        function dijdot = get.dijdot(dp2)
            %returns the derivative of the distance vector [3x1]
            dijdot = dp2.bodyj.rdot + MatOp.B(dp2.bodyj.p, dp2.bodyj.marker{dp2.Qj})*dp2.bodyj.pdot ...
                    -dp2.bodyi.rdot + MatOp.B(dp2.bodyi.p, dp2.bodyi.marker{dp2.Pi})*dp2.bodyi.pdot;
        end  
        
        function aiBar = get.aiBar(dp2)
            % [3x1] local vector
            aiBar = dp2.bodyi.markers{dp2.ai};
        end
       
        function phi = get.phi(dp2) %9.26 slide 14
            %DP2 [1x1] constraint equation
            phi = (dp2.aiBar' * dp2.bodyi.A') * dp2.dij - dp2.f;
        end
        
        function nu = get.nu(dp2) %9.26 slide 12
            nu = dp2.fdot;
        end
        
        function gamma = get.gamma(dp2)% 10.7 slide 8
            %gamma [1x1] is the RHS of the acceleration equation
            Ai      = dp2.bodyi.A*dp2.aiBar; % a = A*abar
            bpdotj  = MatOp.B(dp2.bodyj.pdot, dp2.bodyj.markers{dp2.Qj});
            pdotj   = dp2.bodyj.pdot;
            bpdoti  = MatOp.B(dp2.bodyi.pdot, dp2.bodyi.markers{dp2.Pi});
            pdoti   = dp2.bodyi.pdot;
            bpdotai = MatOp.B(dp2.bodyi.pdot, dp2.aiBar);
            adoti   = MatOp.B(dp2.bodyi.p, dp2.aiBar)*dp2.bodyi.pdot; %aidot = B(p,abar)*pdot
           
            gamma = -Ai'*bpdotj*pdotj + Ai'*bpdoti*pdoi - dp2.dij'*bpdotai*pdoti - 2*adoti'*dp2.dijdot + dp2.fddot;
            
        end
       
        function phi_ri = get.phi_ri(dp2) % s13 lecture 9.28.2016
            %phi_ri [1x3] is the derivative of phi with respect to the 
            %positional coordinates of bodyi. If I is the ground, then
            %phi_ri is [0x0]
            if dp2.bodyi.isground
                phi_ri = [];
            else
                phi_ri = -dp2.aiBar';
            end
        end
        
        function phi_rj = get.phi_rj(dp2) % s13 lecture 9.28.2016
            %phi_ri [1x3] is the derivative of phi with respect to the 
            %positional coordinates of bodyi. If I is the ground, then
            %phi_rj is [0x0]
            if dp2.bodyj.isground
                phi_rj = [];
            else
                phi_rj = dp2.aiBar';
            end
        end
       
        
        function phi_pi = get.phi_pi(dp2) % s13 lecture 9.28.2016
            %phi_pi [1x4] is the derivative of phi with respect to the 
            %orientational coordinates of bodyi. If bodyi is the ground
            %then phi_pi is [0x0]
            if dp2.bodyi.isground
                phi_pi = [];
            else
                phi_pi = dp2.dij*MatOp.B(dp2.bodyi.p,dp2.aiBar) - ...
                         dp2.aiBar'*MatOp.B(dp2.bodyi.p,dp2.bodyi.markers{dp2.Pi}) ;
            end
        end
        
        function phi_pj = get.phi_pj(dp2) % s13 lecture 9.28.2016
            %phi_pi [1x4] is the derivative of phi with respect to the 
            %orientational coordinates of bodyi. If bodyi is the ground
            %then phi_pi is [0x0]
            if dp2.bodyj.isground
                phi_pj = [];
            else
                 phi_pj = dp2.aiBar'*MatOp.B(dp2.bodyj.p,dp2.bodyj.markers{dp1.Qj});
            end
            end
            
    end
end
           
            
    

File: julia/GCons/dp2.jl
#GCons are Geometic Constraints
include("..\\body.jl") #this needs work !!
type dp2
  """
  The dp2 or dot product 2 constraint reflects the fact that the dot product of
  between a vector ai and a second vector PiQj assume a specified value.
  d2 is one of 4 basic GCons that each remove one DOF
  """
  sim::Sim      #reference to the simulation data structures
  bodyi::Body   #bodyi
  bodyj::Body   #bodyj
  Pi::Int       #index of point p on body i (head)
  Qj::Int       #index of point Q on body j (head)
  ai_head::Int  #index of the point at the head of vector ai
  ai_tail::Int  #index of the point at the tial of vector ai
  rDOF::Int     #number of degrees of freedom removed by constraint
  f             #lambda function describing how constraint changes with time
  fdot          #lambda function describing how constraint changes with time
  fddot         #lambda function describing how constraint changes with time



  #constructor function
  function dp2(sim::Sim,bi::Body,bj::Body,Pi::Int,Qj::Int,ai_head::Int, ai_tail = 1, f = t->0 , fdot = t->0, fddot = t->0)
    rDOF = 1; #dp2 removes one DOF
    new(sim,bi,bj,Pi,Qj,ai_head,ai_tail,rDOF,f,fdot,fddot)
  end
end

#----------------begin functions associated with dp1----------------------------
#pseudo - getter methods.
aBari(con::dp2) = pt(con.bodyi,con.ai_head) - pt(con.bodyi,con.ai_tail) #[3x1]
PiQj(con::dp2)  = dij(con.bodyi,cons.bodyj,pt(con.bodyi,con.Pi),pt(con.bodyj,con.Qj))

function ϕ(con::dp2)   #9.26.2016 - slide 14
  """
  constraint equation ϕ
  output: [1 x 1] evaluation of constraint equation value
  """
  phi = aBari(con)'*A(con.bodyi)'PiQj(con) - con.f(con.sim.t)
end

function ν(con::dp2)  #9.26.2016 - slide 15
  """
  RHS of vel equation
  output: [1 x 1] evaluation ν
  """
  nu = con.fdot(con.sim.t)
end

function 	γ(con::dp2)  #10.7.2016 - slide 8
"""
RHS of accel equation
output: [1 x 1] evaluation ν
"""
aiBar = aBari(con) ;
Si = pt(con.bodyi,con.Pi) ; Sj = pt(con.bodyj,con.Qj)
ai = A(con.bodyi)*aiBar
aidot =B(p(con.bodyi) ,aiBar)*pdot(con.bodyi)
pdoti = pdot(con.bodyi) ; pdotj = pdot(con.bodyj)
d_ijdot = dijdot(con.bodyi,con.bodyj,Si,Sj)

gamma = -ai'B(pdotj,Sj)*pdotj + ai'*B(pdoti,Si)*pdoti - PiQj(con)*B(pdoti,aiBar)*pdoti - 2*aidot'd_ijdot + con.fddot(con.sim.t)
end

function ϕ_r(con::dp2)  #9.28.2016 slide 15
  """
  partial derivative of ϕ WRT position position GC's of both bodyi and bodyj
  output: ([1x3],[1x3])
  """
  phi_ri = -aBari(con)'*A(con.bodyi)'
  phi_rj = -phi_ri;
  return phi_ri , ph_rj
end

function ϕ_p(con::dp2)  # #9.28.2016 slide 15
"""
partial derivative of ϕ WRT position orientation GC's of both bodyi and bodyj
output:([1x4],[1x4])
"""
Sj = pt(con.bodyj,con.Qj) ; Si = pt(con.bodyi,con.Pi)  #these are bars
ai = aBari(con)'*A(con.bodyi)
Pj = p(con.bodyj) ; Pi = p(con.bodyi)

phi_pi = PiQj(con)'*B(Pi,aiBar(con)) - ai'B(Pi,Si)
phi_pj = ai'*B(Pj,Sj)

return phi_pi , phi_pj
end
